### ISSUE TITLE
RFC: single() throwing errors

### ISSUE THREAD

**Issue by kiwicopple** [supabase/supabase-js#176]
We have a single() function:

```js
supabase
.from('table')
.select()
.eq('name', 'xx')
.single()
```

If this function returns either 0 rows, or >1 row, then it throws an error

## RFC

Add an option/param to this function `throws` which can have:

- `always` (default) - current behavior
- `never` - catches the error from PostgREST and just returns `{ data: null, error: null }`

this is particularly useful when there are 0 rows. For example in the latest stream: https://youtu.be/p561ogKZ63o?t=2637. In this case I dont really mind if there are no rows - if there was no data it's OK, but it's returning the error too.

The one caveat - what if the query fetches multiple rows? Then the data is `null` and it looks like "no" rows. I don't know how to circumvent this, because the error from PostgREST is the same for both 0 & >1 row.

**Comment by soedirgo** [comment]
That's odd - I thought doing `single()` on 0/many rows just returns PostgREST's error response instead of throwing? This is what I get testing locally:
```js
{
  error: {
    details: 'Results contain 0 rows, application/vnd.pgrst.object+json requires 1 row',
    message: 'JSON object requested, multiple (or no) rows returned'
  },
  data: null,
  count: null,
  status: 406,
  statusText: 'Not Acceptable',
  body: null
}
```

We can distinguish 0 vs. >1 from `error.details`.

**Comment by kiwicopple** [comment]
You're right - it does, however for this use-case I don't want an error at all. Currently the code is:

```js
     let { data, error } = await supabase
        .from('profiles')
        .select(`username, website, avatar_url`)
        .eq('id', user.id)
        .single()

      if (error) { // --------- this is triggered
        throw error
      }
```

And I would like to let this pass:

```js
     let { data, error } = await supabase
        .from('profiles')
        .select(`username, website, avatar_url`)
        .eq('id', user.id)
        .single({  throws: 'never' }) // --------- prevent the error being returned

      if (error) {
        throw error
      }
```

Alternatively, I guess I can do

```js
     let { data, error, status } = await supabase
        .from('profiles')
        .select(`username, website, avatar_url`)
        .eq('id', user.id)
        .single()

      if (error && status !== 406 ) { // --------- could do this
        throw error
      }
```

Although it's a bit harder to grok what's going on. Would love some thoughts. The final option is definitely doable but we will need to document it (wheras the param is essentially self-documenting)

**Comment by phamhieu** [comment]
Maybe we should create a new method singleOrEmpty() and leave single() as it is

**Comment by soedirgo** [comment]
👍

[Slonik](https://github.com/gajus/slonik) uses `one()` analogous to `single()` and `maybeOne()` for 0 or 1.

**Comment by kiwicopple** [comment]
> 0 or 1

@steve-chavez - is there any way to determine that it is 0 or 1 (not many) from the PostgREST response?

**Comment by steve-chavez** [comment]
Also +1 for leaving single() as it is, users have asked about erring for a non-single row response: https://github.com/supabase/supabase/discussions/1606

"I expect an error if authenticated user try to fetch something which is not found"

> Is there any way to determine that it is 0 or 1 (not many) from the PostgREST response?

@kiwicopple Yes, with a combination of the status code + body `error.details`. As Bobbie mentioned [above](https://github.com/supabase/supabase-js/issues/176#issuecomment-845700216), the error message has the number rows in `error.details` as `Results contain <number> rows`. This error message has been stable for some time and it's [tested](https://github.com/PostgREST/postgrest/blob/5c75f0dcc295e6bd847af6d9703fad5b9c3d76c9/test/Feature/SingularSpec.hs#L66-L114).

So for 0, the status code is 406 plus `Results contain 0 rows`. For 1, the status code is 200. Many should be 406 plus `Results contain <x> rows`.

The above plus some internal logic in the library could make a `maybeSingle()` method give `{ data: null, error: null}` or `{ data: {<obj>}, error: null}`.

**Comment by kiwicopple** [comment]
OK great - a new `maybeSingle()` function sound great. Thanks for the input everyone.

It looks like the error in the OP video is returning `multiple (or no) rows` as an error message so I'll dig into the error response and add this function once I have clarity where it's coming from

![image](https://user-images.githubusercontent.com/10214025/119212256-b878c180-bae9-11eb-8ada-e9e17ea482b2.png)
